client/db/bolt/db.go

```diff
@@ -481,7 +481,7 @@ func decodeOrderBucket(oid []byte, oBkt *bbolt.Bucket) (*dexdb.MetaOrder, error)
 			Proof:      *proof,
 			Status:     order.OrderStatus(intCoder.Uint16(oBkt.Get(statusKey))),
 			Host:       string(oBkt.Get(dexKey)),
-			ChangeCoin: oBkt.Get(changeKey),
+			ChangeCoin: oBkt.Get(changeKey), // what if it was stored as nil?
```

Comment:
Good point. Should be storing nil interface. Should also favor `len(change) == 0` over `nil` check throughout.

Reply:
Since `ChangeCoin` is a `order.CoinID`, which is a `[]byte`, I guess I'm pointing out here that `oBkt.Get(changeKey)` may or may not return nil vs []byte{}... just need to check and see since I'm not that familiar with bbolt.  Maybe this is just fine and `ChangeCoin` will come back `nil`... dunno.

Reply:
Just confirmed that putting `nil` comes back as `[]byte{}` instead. 84db38c5  The `len(change) == 0` checks in higher level code is also a good idea that I'll get on as well.

Reply:
Ah. Yep, **bbolt** will return `nil` for `nil` and `[]byte{}` for `[]byte{}`.

Reply:
Pushed a commit to do length checks in favor of nil checks anyway.  e2e690809a5e1ec968f529c5489f8559ac2fdcb4

But I observed that bolt returns `[]byte{}` always, even when `nil` is stored.  That is, in https://github.com/decred/dcrdex/commit/84db38c5ad6d07f7342d89993cc54529ebb05cc9, if you get rid of that `if len(changeCoin) == 0` I added in `decodeOrderBucket`, it would never set the returned `dexdb.OrderMetaData.ChangeCoin` to `nil` even if you've stored it that way.

Reply:
Hmmm. My test revealed otherwise.

```go
func TestThis(t *testing.T) {
	boltdb := newTestDB(t)
	bktID := []byte("abc")
	k := []byte("k")
	err := boltdb.Update(func(tx *bbolt.Tx) error {
		bkt, err := tx.CreateBucketIfNotExists(bktID)
		if err != nil {
			return err
		}
		err = bkt.Put(k, []byte{})
		if err != nil {
			return err
		}
		b := bkt.Get(k)
		fmt.Println("[]byte{} is nil?", b == nil)
		err = bkt.Put(k, nil)
		if err != nil {
			return err
		}
		b = bkt.Get(k)
		fmt.Println("nil is nil?", b == nil)
		return nil
	})
	if err != nil {
		t.Fatalf("Update error: %v", err)
	}
}
```

Reply:
Oh, not good. That means `(*BoltDB).Get` is broken for `nil`.

But I am not seeing why the `bkt.Get` from `decodeOrderBucket` will not return nil, only `[]byte{}`.

Reply:
Well, actually we seemed to have stumbled on some undefined behavior in bolt.  For a stored nil slice, you will receive either an empty non-nil slice or a nil slice, depending on whether the `Cursor` is retrieve the value from the "page" or "node", respectively.

https://github.com/etcd-io/bbolt/blob/232d8fc87f50244f9c808f4745759e08a304c029/cursor.go#L334-L351

In the "retrieve value from page" branch, it uses [`(*leafPageElement).value`](https://github.com/etcd-io/bbolt/blob/232d8fc87f50244f9c808f4745759e08a304c029/page.go#L129-L134) > `unsafeByteSlice` to get the value byte slice, which always results in a non-nil slice:

https://github.com/etcd-io/bbolt/blob/232d8fc87f50244f9c808f4745759e08a304c029/unsafe.go#L27

This is all it can do because a `leafPageElement` only knows the "size" and 0 does not disambiguate nil from empty non-nil slice.  https://github.com/etcd-io/bbolt/blob/232d8fc87f50244f9c808f4745759e08a304c029/page.go#L115-L120

I don't see a fix for bolt, other than possibly to revise their use of `inode` to always return `[]byte{}` when `len(inode.value)` is zero rather than returning whatever was stored there (`nil` or `[]byte{}`).

I'll make a minimal repro to demonstrate, but it seems simple to hit with enough entries in the same bucket.

Reply:
With orders:

```go

func TestOrderChange(t *testing.T) {
	boltdb := newTestDB(t)
	// Create an account to use.
	acct := dbtest.RandomAccountInfo()
	err := boltdb.CreateAccount(acct)
	if err != nil {
		t.Fatalf("CreateAccount error: %v", err)
	}

	ord := randOrderForMarket(randU32(), randU32())
	mordIn := &db.MetaOrder{
		MetaData: &db.OrderMetaData{
			Status: 3,
			Host:   acct.Host,
			Proof:  db.OrderProof{DEXSig: randBytes(73)},
		},
		Order: ord,
	}

	fmt.Printf("%#v\n", mordIn.MetaData.ChangeCoin)
	fmt.Printf("%#v\n", mordIn.MetaData.Proof.Preimage)

	err = boltdb.UpdateOrder(mordIn)
	if err != nil {
		t.Fatalf("error inserting order: %v", err)
	}

	mord, err := boltdb.Order(ord.ID())
	if err != nil {
		t.Fatalf("unable to retrieve order by id")
	}
	fmt.Printf("%#v\n", mord.MetaData.ChangeCoin)     // incorrect
	fmt.Printf("%#v\n", mord.MetaData.Proof.Preimage) // correct
	if mord.MetaData.ChangeCoin != nil {
		t.Errorf("ChangeCoin was not nil: %#v", mord.MetaData.ChangeCoin)
	}
}
```

Just plain bolt:

```go
func TestNilValue(t *testing.T) {
	tDir, _ := ioutil.TempDir("", "dbtest")
	defer os.RemoveAll(tDir)
	tDbFile := filepath.Join(tDir, "bolt.db")
	boltdb, err := bbolt.Open(tDbFile, 0600, nil)
	if err != nil {
		t.Fatal(err)
	}

	bktID := []byte("bucket")
	nk := []byte("nilkey")
	ek := []byte("emptykey")

	err = boltdb.Update(func(tx *bbolt.Tx) error {
		bkt, err := tx.CreateBucketIfNotExists(bktID)
		if err != nil {
			return err
		}

		err = bkt.Put(nk, nil)
		if err != nil {
			return err
		}

		// (*Cursor).keyValue retrieves value from node, returning the nil byte
		// slice from inode.value.
		b := bkt.Get(nk)
		if b != nil {
			return fmt.Errorf("not nil, %#v", b)
		}

		err = bkt.Put(ek, []byte{})
		if err == nil {
			return err
		}

		b = bkt.Get(ek)
		if b == nil {
			return fmt.Errorf("not []byte{}, %#v", b)
		}

		return nil
	})
	// Commit "spills" to pages.
	if err != nil {
		t.Fatalf("Update error: %v", err)
	}

	err = boltdb.View(func(tx *bbolt.Tx) error {
		bkt := tx.Bucket(bktID)
		// (*Cursor).keyValue retrieves value from page via unsafeByteSlice,
		// always creating a non-nil byte slice.
		b := bkt.Get(nk)
		if b != nil {
			// Fails here.
			t.Errorf("not nil, %#v", b)
		}
		b = bkt.Get(ek)
		if b == nil {
			// OK here.
			t.Errorf("not []byte{}, %#v", b)
		}
		return nil
	})
	if err != nil {
		t.Fatalf("View error: %v", err)
	}
}
```